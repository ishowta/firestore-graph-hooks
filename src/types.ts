import {
  CollectionReference,
  DocumentReference,
  QuerySnapshot,
  DocumentData,
  Query,
  SnapshotMetadata,
} from 'firebase/firestore';
import { PickOptional, SelectiveOptional, Expand } from './utils';

export type GraphQueryDocumentSnapshot<T extends DocumentData> = {
  data: T;
  exist: true;
  id: string;
  ref: DocumentReference<T>;
  metadata: SnapshotMetadata;
};

export type GraphDocumentSnapshot<T extends DocumentData> =
  | GraphQueryDocumentSnapshot<T>
  | {
      data: undefined;
      exist: false;
      id: string;
      ref: DocumentReference<T>;
      metadata: SnapshotMetadata;
    };

/**
 * FirestoreのID
 */
export type ID = string & {};

/**
 * 得られるCollectionに付属するメタデータ
 */
export type CollectionMetadata<T> = {
  __snapshot__: QuerySnapshot<T>;
};

export type WithCollectionMetadata<T extends DocumentData> =
  CollectionMetadata<T> & GraphQueryDocumentSnapshot<T>[];

export type AnyReference = DocumentReference | Query;

/**
 * ドキュメントのフィールドからリファレンスのフィールドのみを取り出す
 *
 * 元のドキュメントとの互換性を保つため`${string}Ref`の形のフィールドのみ使えるようにする
 */
type PickRefField<T extends DocumentData> = keyof {
  [K in keyof T as K extends `${string}Ref` ? K : never]: NonNullable<
    T[K]
  > extends AnyReference
    ? K
    : never;
};

/**
 * クエリの型
 *
 * TODO: 否定形や存在型が無いのでextra fieldを`unknown`にしてしまっているが他に方法は無い？
 */
export type GraphQuery<T extends DocumentData> =
  // ref fieldを含んだクエリ（extra fieldも入れられる）
  | (({
      [K in PickRefField<T>]?: NonNullable<T[K]> extends
        | DocumentReference<infer U>
        | CollectionReference<infer U>
        ? U extends DocumentData
          ? GraphQuery<U>
          : never
        : never;
    } & {
      [K in Exclude<keyof T, PickRefField<T>>]?: never;
    }) & {
      [K in string]: unknown | [AnyReference, unknown, boolean];
    })
  // extra fieldのみのクエリ
  | ({ [K in keyof T]?: never } & {
      [K in string]: unknown | [AnyReference, unknown, boolean];
    })
  // ドキュメントを引数にとってクエリを返す関数
  | ((data: GraphDocumentSnapshot<T>) => GraphQuery<T>);

type GraphQueryQueryType<T, Q extends GraphQuery<T>> = Q extends (
  ...args: any
) => any
  ? ReturnType<Q>
  : Q;

type RequiredGraphQuery<Q> = Q extends (...args: any) => any ? Q : Required<Q>;

export type JoinedDataInner<
  T extends DocumentData,
  Q extends GraphQuery<T>
> = SelectiveOptional<
  {
    /**
     * ドキュメントのもともとのフィールド
     */
    [K in Exclude<keyof T, keyof GraphQueryQueryType<T, Q>>]: T[K];
  } & {
    /**
     * クエリで指定されたリファレンスフィールド
     */
    [K in keyof T &
      keyof GraphQueryQueryType<T, Q> as K extends `${infer OriginalK}Ref`
      ? OriginalK
      : K]: RefToDoc<NonNullable<T[K]>> extends DocumentData
      ? RequiredGraphQuery<GraphQueryQueryType<T, Q>[K]> extends GraphQuery<
          RefToDoc<NonNullable<T[K]>>
        >
        ?
            | JoinedData<
                T[K],
                RequiredGraphQuery<GraphQueryQueryType<T, Q>[K]>,
                false
              >
            | (null extends T[K] ? null : never)
        : never
      : never;
  } & {
    /**
     * クエリで追加されたエクストラフィールド
     */
    [K in Exclude<
      keyof GraphQueryQueryType<T, Q>,
      keyof T
    >]: GraphQueryQueryType<T, Q>[K] extends
      | [infer Ref, infer UQuery, infer GuaranteedToExist]
      | undefined
      ? GuaranteedToExist extends boolean
        ? Ref extends AnyReference
          ? UQuery extends Function
            ? UQuery extends GraphQuery<RefToDoc<Ref>>
              ? JoinedData<Ref, UQuery, GuaranteedToExist>
              : never
            : Required<UQuery> extends GraphQuery<RefToDoc<Ref>>
            ? JoinedData<Ref, Required<UQuery>, GuaranteedToExist>
            : never
          : never
        : never
      : never;
  },
  keyof PickOptional<T> | keyof PickOptional<GraphQueryQueryType<T, Q>>
>;

export type RefToDoc<R extends AnyReference> = R extends
  | DocumentReference<infer D>
  | undefined
  ? D
  : R extends Query<infer D> | undefined
  ? D
  : never;

export type JoinedData<
  R extends AnyReference,
  Q extends GraphQuery<RefToDoc<R>>,
  GuaranteedToExist extends boolean
> = R extends DocumentReference<infer U>
  ? Q extends GraphQuery<U>
    ? {} extends Omit<JoinedDataInner<U, Q>, keyof U>
      ? true extends GuaranteedToExist
        ? GraphQueryDocumentSnapshot<U>
        : GraphDocumentSnapshot<U>
      : (true extends GuaranteedToExist
          ? GraphQueryDocumentSnapshot<U>
          : GraphDocumentSnapshot<U>) & {
          data: Expand<Omit<JoinedDataInner<U, Q>, keyof U>>;
        }
    : never
  : R extends CollectionReference<infer U>
  ? Q extends GraphQuery<U>
    ? {} extends Omit<JoinedDataInner<U, Q>, keyof U>
      ? WithCollectionMetadata<U>
      : (GraphQueryDocumentSnapshot<U> & {
          data: Expand<Omit<JoinedDataInner<U, Q>, keyof U>>;
        })[] &
          CollectionMetadata<U>
    : never
  : never;
