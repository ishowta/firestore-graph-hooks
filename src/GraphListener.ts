import {
  DocumentReference,
  FirestoreError,
  Query,
  onSnapshot,
  Unsubscribe,
} from 'firebase/firestore';
import { GraphDocumentSnapshot, GraphQuery } from './types';
import {
  makeGraphDocumentSnapshot,
  makeGraphQueryDocumentSnapshot,
} from './converter';
import { getObjectLogger, insert } from './utils';
import { GraphQueryListener } from './GraphQueryListener';
import { Logger } from 'loglevel';

export interface GraphListener {
  ref: any;
  query: any;
  loading: boolean;

  /**
   * クエリに更新があるかどうかを返す
   */
  updateQuery(newQuery: GraphQuery<any>, dryRun: boolean): boolean;

  /**
   * サブクエリの購読を止める
   */
  unsubscribe(): void;
}

export const makeGraphListener = (
  ref: DocumentReference | Query,
  query: any,
  handleUpdate: (result: any) => void,
  handleError: (error: FirestoreError) => void
): GraphListener => {
  if (ref instanceof Query) {
    return new GraphCollectionListener(ref, query, handleUpdate, handleError);
  } else {
    return new GraphDocumentListener(ref, query, handleUpdate, handleError);
  }
};

export class GraphDocumentListener implements GraphListener {
  ref: DocumentReference;
  listenerUnsubscriber: Unsubscribe;
  query: any;
  queryListener: GraphQueryListener | undefined;
  loading: boolean;
  isSnapshotInitialized: boolean;
  logger: Logger;

  constructor(
    ref: DocumentReference,
    query: any,
    handleUpdate: (result: any) => void,
    handleError: (error: FirestoreError) => void
  ) {
    this.logger = getObjectLogger(this, ref.path);
    this.loading = true;
    this.isSnapshotInitialized = false;
    this.ref = ref;
    this.query = query;
    this.queryListener = undefined;

    const onUpdate = (result: any) => {
      this.logger.debug('onUpdate', result);
      this.loading = false;
      handleUpdate(result);
    };

    this.logger.debug('init', (ref as any).path);

    this.listenerUnsubscriber = onSnapshot(ref, (rawSnapshot) => {
      const snapshot = makeGraphDocumentSnapshot(rawSnapshot);
      this.logger.debug('onSnapshot', snapshot);
      if (this.queryListener) {
        if (snapshot.exist) {
          this.queryListener.updateSnapshot(snapshot, false);
          onUpdate(this.queryListener.result);
        } else {
          onUpdate(undefined);
        }
      } else {
        if (snapshot.exist) {
          this.queryListener = new GraphQueryListener(
            snapshot,
            query,
            onUpdate
          );
        } else {
          onUpdate(undefined);
          this.loading = false;
        }
      }
      this.isSnapshotInitialized = true;
    });
  }

  updateQuery(newQuery: any, dryRun: boolean): boolean {
    this.logger.debug('updateQuery', newQuery, dryRun);
    if (!dryRun) {
      this.query = newQuery;
    }
    if (this.queryListener) {
      return this.queryListener.updateQuery(newQuery, dryRun);
    } else {
      return false;
    }
  }

  unsubscribe(): void {
    this.logger.debug('unsubscribe');
    if (this.queryListener) {
      this.queryListener.unsubscribe();
    }
    this.listenerUnsubscriber();
  }
}

export class GraphCollectionListener implements GraphListener {
  ref: Query;
  listenerUnsubscriber: Unsubscribe;
  query: any;
  queryListeners: Record<string, GraphQueryListener>;
  result: GraphDocumentSnapshot<any>[];
  loading: boolean;
  logger: Logger;

  constructor(
    ref: Query,
    query: any,
    handleUpdate: (result: any) => void,
    handleError: (error: FirestoreError) => void
  ) {
    this.logger = getObjectLogger(this, (ref as any).path);
    this.loading = true;
    this.ref = ref;
    this.query = query;
    this.result = [];
    this.queryListeners = {};

    const onUpdate = () => {
      this.logger.debug('onUpdate');

      // FIXME: initializedではなくloadingを見るべきでは?
      this.logger.debug(
        `${
          Object.values(this.queryListeners).filter(
            (queryListener) => queryListener.isQueryInitialized
          ).length
        }/${this.result.length} initialized`
      );

      if (
        Object.keys(this.queryListeners).length === this.result.length &&
        Object.values(this.queryListeners).every(
          (queryListener) => queryListener.isQueryInitialized
        )
      ) {
        this.logger.debug('updated');
        this.loading = false;
        handleUpdate(this.result);
      }
    };

    const onUpdateWithResult = (
      path: string,
      result: GraphDocumentSnapshot<any>
    ) => {
      this.logger.debug('onUpdateWithResult', path, result);
      const docIndex = this.result.findIndex((res) => res.ref.path === path);
      if (docIndex === -1) {
        this.logger.debug('path not found, skip update.', this.result, path);
        return;
      }

      this.result[docIndex] = result;
      this.result = [...this.result];

      onUpdate();
    };

    this.logger.debug('init', (ref as any).path);

    this.listenerUnsubscriber = onSnapshot(ref, (querySnapshot) => {
      this.logger.debug('onSnapshot', querySnapshot);

      // set result
      for (const docChange of querySnapshot.docChanges()) {
        const snapshot = makeGraphQueryDocumentSnapshot(docChange.doc);
        switch (docChange.type) {
          case 'added': {
            this.result = insert(this.result, snapshot, docChange.newIndex);
            break;
          }
          case 'removed':
            this.result.splice(docChange.oldIndex, 1);
            break;
          case 'modified':
            if (docChange.oldIndex !== docChange.newIndex) {
              const temp = this.result.splice(docChange.oldIndex, 1)[0];
              this.result = insert(this.result, temp, docChange.newIndex);
            }
            break;
        }
      }

      // update snapshot
      for (const docChange of querySnapshot.docChanges()) {
        this.logger.debug('docChange', docChange.type);
        const snapshot = makeGraphQueryDocumentSnapshot(docChange.doc);
        switch (docChange.type) {
          case 'added': {
            this.queryListeners[docChange.doc.ref.path] =
              new GraphQueryListener(snapshot, query, (result) =>
                onUpdateWithResult(docChange.doc.ref.path, result)
              );
            break;
          }
          case 'removed':
            this.queryListeners[docChange.doc.ref.path].unsubscribe();
            break;
          case 'modified':
            this.queryListeners[docChange.doc.ref.path].updateSnapshot(
              snapshot,
              false
            );
            onUpdateWithResult(
              docChange.doc.ref.path,
              this.queryListeners[docChange.doc.ref.path].result
            );
            break;
        }
      }
      onUpdate();
    });
  }

  updateQuery(newQuery: any, dryRun: boolean): boolean {
    this.logger.debug('updateQuery', newQuery, dryRun);
    if (!dryRun) {
      this.query = newQuery;
    }
    if (this.queryListeners) {
      const hasUpdate = Object.values(this.queryListeners).some(
        (queryListener) => queryListener.updateQuery(newQuery, dryRun)
      );
      if (dryRun) return hasUpdate;
      if (hasUpdate) {
        this.loading = true;
      }
      return hasUpdate;
    } else {
      return false;
    }
  }

  unsubscribe(): void {
    this.logger.debug('unsubscribe');
    if (this.queryListeners) {
      Object.values(this.queryListeners).forEach((queryListener) =>
        queryListener.unsubscribe()
      );
    }
    this.listenerUnsubscriber();
  }
}
