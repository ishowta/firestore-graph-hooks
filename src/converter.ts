import { DocumentSnapshot, QueryDocumentSnapshot } from 'firebase/firestore';
import {
  AnyReference,
  DocumentData,
  GetQueryType,
  GraphDocumentSnapshot,
  GraphQuery,
  GraphQueryDocumentSnapshot,
  GraphQueryExtensionField,
  GraphQueryGenerator,
} from './types';

export const makeGraphQueryDocumentSnapshot = <T extends DocumentData>(
  snapshot: QueryDocumentSnapshot<T>
): GraphQueryDocumentSnapshot<T> => {
  return {
    exist: true,
    data: snapshot.data(),
    id: snapshot.id,
    ref: snapshot.ref,
    metadata: snapshot.metadata,
  };
};

export const makeGraphDocumentSnapshot = <T extends DocumentData>(
  snapshot: DocumentSnapshot<T>
): GraphDocumentSnapshot<T> => {
  if (snapshot.exists()) {
    return {
      exist: true,
      data: snapshot.data(),
      id: snapshot.id,
      ref: snapshot.ref,
      metadata: snapshot.metadata,
    };
  } else {
    return {
      exist: false,
      data: undefined,
      id: snapshot.id,
      ref: snapshot.ref,
      metadata: snapshot.metadata,
    } as GraphDocumentSnapshot<T>;
  }
};

export type SubQueryData<T extends DocumentData, Ref extends AnyReference<T>> =
  | {
      type: 'external';
      key: `${string}Ref`;
      ref: AnyReference<T> | null | undefined;
      query: GraphQuery<T> | GraphQueryGenerator<Ref>;
    }
  | {
      type: 'extension';
      key: string;
      ref: AnyReference<T> | undefined;
      query: GraphQuery<T> | GraphQueryGenerator<Ref>;
    };

export const makeQuery = <
  T extends DocumentData,
  Ref extends AnyReference<T>,
  Q extends GraphQuery<T> | GraphQueryGenerator<Ref>
>(
  snapshot: GraphQueryDocumentSnapshot<T>,
  queryGenerator: Q
) => {
  return (
    typeof queryGenerator === 'function'
      ? queryGenerator(snapshot)
      : queryGenerator
  ) as GetQueryType<T, Ref, Q>;
};

export const isQueryKey = (
  snapshotKey: string
): snapshotKey is `${string}Ref` => {
  return snapshotKey.endsWith('Ref');
};

export const makeSubQuery = <
  T extends DocumentData,
  U extends DocumentData,
  URef extends AnyReference<U>,
  Q extends GraphQuery<U> | GraphQueryGenerator<URef> | GraphQueryExtensionField
>(
  snapshot: GraphQueryDocumentSnapshot<T>,
  subQueryKey: string,
  rawSubQuery: Q
): SubQueryData<U, URef> => {
  // フィールドがオプショナルフィールドである場合、実態が無いので拡張フィールドなのか外部キーなのかの区別がつかない。
  // 仕方がないので拡張キーは`[ref, query]`の形式であるとしてそれで判断する
  // documentKeys.includes(subSubQueryKey) &&
  const type = Array.isArray(rawSubQuery) ? 'extension' : 'external';
  switch (type) {
    case 'external': {
      const query = rawSubQuery as GraphQuery<U> | GraphQueryGenerator<URef>;
      if (isQueryKey(subQueryKey)) {
        return {
          key: subQueryKey,
          type: 'external',
          ref: snapshot.data[subQueryKey] as AnyReference<U> | null | undefined,
          query: query,
        };
      } else {
        throw new Error('Required Ref');
      }
    }

    case 'extension': {
      const [ref, query] = rawSubQuery as GraphQueryExtensionField;
      return {
        key: subQueryKey,
        type: 'extension',
        ref: ref as AnyReference<U>,
        query: query as GraphQuery<U>,
      };
    }
  }
};
