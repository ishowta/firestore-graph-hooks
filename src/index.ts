import loglevel from 'loglevel';
import { apply, reg } from 'loglevel-plugin-prefix';

loglevel.setLevel('info');
reg(loglevel);
apply(loglevel, {
  format(level, name, timestamp) {
    return `[${timestamp}] ${level} ${name}:`;
  },
});

export const logger = loglevel;

export * from './useQuery';
