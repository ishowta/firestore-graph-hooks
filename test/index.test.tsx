import { generateSeed } from './seed';
import { field } from '../';
import { getKanbans, getProjects, getTodoLists } from './schema';
import { getApp } from './firebase';
import {
  DocumentReference,
  Firestore,
  orderBy,
  query,
  Query,
  Timestamp,
} from 'firebase/firestore';
import { makeGraphListener } from '../src/GraphListener';
import Channel from '@nodeguy/channel';
import traverse from 'traverse';

let firestore: Firestore;

beforeAll(async () => {
  firestore = (await getApp()).firestore as any;
  await generateSeed(firestore);
});

test('makeGraphListener', async () => {
  const queryChannel = Channel();

  makeGraphListener(
    query(getProjects(firestore), orderBy('createdAt')),
    (project: any) => ({
      ownerRef: {
        nowPlayingRef: {},
      },
      kanbans: field(
        query(getKanbans(project.ref), orderBy('createdAt')),
        (kanban) => ({
          todoLists: field(
            query(getTodoLists(kanban.ref), orderBy('createdAt')),
            {}
          ),
        })
      ),
    }),
    (result) => {
      queryChannel.push(result);
    },
    () => {}
  );

  const res = await queryChannel.shift();

  expect(
    traverse(res).map(function (x) {
      if (x instanceof DocumentReference || x instanceof Query)
        this.update('[Ref]');
      if (this.key === 'id') this.update('[ID]');
      if (x instanceof Timestamp) this.update('[Timestamp]');
    })
  ).toMatchSnapshot();
});
