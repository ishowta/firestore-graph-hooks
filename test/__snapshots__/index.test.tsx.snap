// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeGraphListener 1`] = `
Array [
  Object {
    "data": Object {
      "createdAt": "[Timestamp]",
      "kanbans": Array [
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "0th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "1th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "2th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
      ],
      "owner": Object {
        "data": Object {
          "createdAt": "[Timestamp]",
          "name": "0man",
          "nowPlaying": Object {
            "data": Object {
              "createdAt": "[Timestamp]",
              "scheduledTime": 0,
              "title": "0th todo",
            },
            "exist": true,
            "id": "[ID]",
            "metadata": SnapshotMetadata {
              "fromCache": false,
              "hasPendingWrites": false,
            },
            "ref": "[Ref]",
          },
          "nowPlayingRef": "[Ref]",
        },
        "exist": true,
        "id": "[ID]",
        "metadata": SnapshotMetadata {
          "fromCache": false,
          "hasPendingWrites": false,
        },
        "ref": "[Ref]",
      },
      "ownerRef": "[Ref]",
      "title": "0th project",
    },
    "exist": true,
    "id": "[ID]",
    "metadata": SnapshotMetadata {
      "fromCache": false,
      "hasPendingWrites": false,
    },
    "ref": "[Ref]",
  },
  Object {
    "data": Object {
      "createdAt": "[Timestamp]",
      "kanbans": Array [
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "0th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "1th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "2th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
      ],
      "owner": Object {
        "data": Object {
          "createdAt": "[Timestamp]",
          "name": "1man",
          "nowPlaying": Object {
            "data": Object {
              "createdAt": "[Timestamp]",
              "scheduledTime": 0,
              "title": "0th todo",
            },
            "exist": true,
            "id": "[ID]",
            "metadata": SnapshotMetadata {
              "fromCache": false,
              "hasPendingWrites": false,
            },
            "ref": "[Ref]",
          },
          "nowPlayingRef": "[Ref]",
        },
        "exist": true,
        "id": "[ID]",
        "metadata": SnapshotMetadata {
          "fromCache": false,
          "hasPendingWrites": false,
        },
        "ref": "[Ref]",
      },
      "ownerRef": "[Ref]",
      "title": "1th project",
    },
    "exist": true,
    "id": "[ID]",
    "metadata": SnapshotMetadata {
      "fromCache": false,
      "hasPendingWrites": false,
    },
    "ref": "[Ref]",
  },
  Object {
    "data": Object {
      "createdAt": "[Timestamp]",
      "kanbans": Array [
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "0th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "1th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
        Object {
          "data": Object {
            "createdAt": "[Timestamp]",
            "title": "2th kanban",
            "todoLists": Array [
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 0 World!",
                  "name": "0th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 1 World!",
                  "name": "1th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
              Object {
                "data": Object {
                  "createdAt": "[Timestamp]",
                  "memo": "Hello 2 World!",
                  "name": "2th todoList",
                },
                "exist": true,
                "id": "[ID]",
                "metadata": SnapshotMetadata {
                  "fromCache": false,
                  "hasPendingWrites": false,
                },
                "ref": "[Ref]",
              },
            ],
          },
          "exist": true,
          "id": "[ID]",
          "metadata": SnapshotMetadata {
            "fromCache": false,
            "hasPendingWrites": false,
          },
          "ref": "[Ref]",
        },
      ],
      "owner": Object {
        "data": Object {
          "createdAt": "[Timestamp]",
          "name": "2man",
          "nowPlaying": Object {
            "data": Object {
              "createdAt": "[Timestamp]",
              "scheduledTime": 0,
              "title": "0th todo",
            },
            "exist": true,
            "id": "[ID]",
            "metadata": SnapshotMetadata {
              "fromCache": false,
              "hasPendingWrites": false,
            },
            "ref": "[Ref]",
          },
          "nowPlayingRef": "[Ref]",
        },
        "exist": true,
        "id": "[ID]",
        "metadata": SnapshotMetadata {
          "fromCache": false,
          "hasPendingWrites": false,
        },
        "ref": "[Ref]",
      },
      "ownerRef": "[Ref]",
      "title": "2th project",
    },
    "exist": true,
    "id": "[ID]",
    "metadata": SnapshotMetadata {
      "fromCache": false,
      "hasPendingWrites": false,
    },
    "ref": "[Ref]",
  },
]
`;
